### PR TITLE
Refactor server_dap to use command-registration style

### PR DESCRIPTION
**Pull request by andyw8**
Best viewed with whitespace ignored: https://github.com/Shopify/debug/pull/12/files?w=1

## Description

Closes https://github.com/Shopify/ruby-dev-exp-issues/issues/840

### CODE REVIEW

File: lib/debug/server_dap.rb [hunk 1]
```diff
@@ -274,188 +274,233 @@ def recv_request
       retry
     end
 
-    def process
-      while req = recv_request
-        raise "not a request: #{req.inspect}" unless req['type'] == 'request'
-        args = req.dig('arguments')
-
-        case req['command']
+    class RequestHandler
```
Comment (andyw8) [inline lib/debug/server_dap.rb:277]: (using a class to hold only a single field may be unnecessary)

File: lib/debug/server_dap.rb [hunk 2]
```diff
@@ -64,6 +65,14 @@ def self.setup debug_port
       end
     end
 
+    def self.register_request *names, &b
```
Comment (andyw8) [inline lib/debug/server_dap.rb:68]: The style/naming is not what I'd typically use, but I've tried keep it consistent with what's already in use.

File: lib/debug/server_dap.rb (hunk 1 above)
Comment (vinistock) [reply lib/debug/server_dap.rb:286]: How do you envision us adding metrics through this class?

Comment (andyw8) [reply lib/debug/server_dap.rb:286]: > How do you envision us adding metrics through this class?

One approach could be to add a command line option, e,g. `--request-handler="Shopify::RequestHandler"`, allowing a custom class to provided. 

Maybe @st0012 already has something else in mind though.

Comment (vinistock) [reply lib/debug/server_dap.rb:286]: We may not even need this class at all. We can override register_request
```ruby
def self.register_request *names, &b
      instrumented = -> { 
        measure_stuff do
          b.call
        end
      }
      names.each { |name| REGISTERED_REQUESTS[name] = instrumented }
end
```

Comment (andyw8) [reply lib/debug/server_dap.rb:286]: Yeah that would also work, but I worry it may not be resilient to future changes within `debug`.

**Pull request closed without merging**